### PR TITLE
Logicandcompare

### PR DESCRIPTION
**Pull request by franksl**
This builds up on my previous pull request (https://github.com/neonxp/MathExecutor/pull/49) adding logical and compare operators <, <=, >, >=, ==, !=, ||, &&

### CODE REVIEW

**Comment by phpfui** [comment]
While I fundamentally agree with the idea of adding logical operators to the package, there are a couple of issues that we need to resolve before we can merge.  We also need to update the docs.  I can do that before the release, but we need to fix the operator precedence issues and a couple of minor text messages.  Will try to put comments in the correct locations, but still new to github code reviews.

Also, Alex, do you want to weigh in on the idea of adding comparison operators to the package?

Basically conditional operators should return an int of 1 (true) or 0 (false).  But this is only really useful if we have a comparison type function.  I would propose we add a built in if function that would be look like this:

if ($condition, $trueReturnValue, $falseReturnValue)

You would then be able do do something like if ($a >= $b, $b / $a, $a - $b)

Also noticing other packages like this (this one still being the best), and the others implement a much wider variety of built in functions.  Thinking conditionals and expanded function support is version 1.1.

And Franco, thanks for the work on this.

File: src/NXP/Classes/Token/TokenAnd.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @throws \NXP\Exception\IncorrectExpressionException
+     */
+    public function execute(&$stack)
+    {
+        $op2 = array_pop($stack);
+        $op1 = array_pop($stack);
+
+        if ($op1 === null || $op2 === null) {
+            throw new IncorrectExpressionException("And requires two operators");
```
Comment (phpfui) [inline src/NXP/Classes/Token/TokenAnd.php]: Change "And" here to && to be consistent with the others.

File: src/NXP/Classes/Token/TokenOr.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @throws \NXP\Exception\IncorrectExpressionException
+     */
+    public function execute(&$stack)
+    {
+        $op2 = array_pop($stack);
+        $op1 = array_pop($stack);
+
+        if ($op1 === null || $op2 === null) {
+            throw new IncorrectExpressionException("Or requires two operators");
```
Comment (phpfui) [inline src/NXP/Classes/Token/TokenOr.php]: Change "Or" here to ||

File: tests/MathTest.php [hunk 3]
```diff
@@ -110,6 +110,23 @@ public function providerExpressions()
...
+            ['5 != 5'],
+            ['5 > 3'],
+            ['3 > 5'],
+            ['3 >= 5'],
+            ['3 >= 3'],
+            ['3 < 5'],
+            ['5 < 3'],
+            ['3 <= 5'],
+            ['5 <= 5'],
         ];
```
Comment (phpfui) [inline tests/MathTest.php]: Try adding the following tests:

            ['1 + 5 == 3 + 1'],
            ['1 + 5 == 5 + 1'],
            ['1 + 5 != 3 + 1'],
            ['1 + 5 != 5 + 1'],
            ['1 + 5 > 3 + 1'],
            ['1 + 3 > 5 + 1'],
            ['1 + 3 >= 5 + 1'],
            ['1 + 3 >= 3 + 1'],
            ['1 + 3 < 5 + 1'],
            ['1 + 5 < 3 + 1'],
            ['1 + 3 <= 5 + 1'],
            ['1 + 5 <= 5 + 1'],

I think you will find you will need to fix operator precedence.

Also, I would add additional tests with multiply and divide.  Their precedence is different, so we should make sure our precedence is right.  I would also test () in the same fashion.  Our library should match PHP operator precedence and the best thing about how Alex set up the tests is that we are comparing against PHP itself. Always better to have more tests.

**Comment by franksl** [comment]
Hi,
The if() function is exactly what I needed and the work in this PR is the preparation for the if() function. I've already implemented it in an extension that I use in my code but I can publish it as a PR. I'll look into the suggested changes as soon as possible.

**Comment by phpfui** [comment]
Yes, please add if function in next commit.  Thanks!  This makes no sense without it actually.

Will be a very nice enhancement to the package.

**Comment by franksl** [comment]
Hi,
I solved the issues (I hope!) and added the if function. I put some comments in the commit messages that should explain how I worked.
I also solved a small problem with cache key creation.
Hope everything is good now,
Thanks,
Franco

**Comment by phpfui** [comment]
Looks good!  I'll update the readme and issue a release at some point.  Thanks for the PR!

**Comment by phpfui** [comment]
V1.1.0 is released.  I did change the name of TokenUnequal to TokenNotEqual, but should not make a difference to anyone.

**Pull request merged**